sixtyfps_compiler/passes/lower_layout.rs

```diff
@@ -49,246 +123,185 @@ fn lower_grid_layout(
+}
+
+impl GridLayout {
+    fn add_element(
+        &mut self,
+        item_element: &ElementRc,
+        (row, col): (&mut u16, &mut u16),
+        layout_cache_prop: &NamedReference,
+        diag: &mut BuildDiagnostics,
+    ) {
+        let mut get_const_value = |name: &str| {
+            item_element
+                .borrow_mut()
+                .bindings
+                .remove(name)
+                .and_then(|e| eval_const_expr(&e.expression, name, &e, diag))
+        };
+        let colspan = get_const_value("colspan").unwrap_or(1);
+        let rowspan = get_const_value("rowspan").unwrap_or(1);
+        if let Some(r) = get_const_value("row") {
+            *row = r;
+            *col = 0;
+        }
+        if let Some(c) = get_const_value("col") {
+            *col = c;
+        }
+
+        let index = self.elems.len();
+        if let Some(layout_item) = create_layout_item(&item_element, index, layout_cache_prop, diag)
+        {
+            self.elems.push(GridLayoutElement {
+                col: *col,
+                row: *row,
+                colspan,
+                rowspan,
+                item: layout_item,
+            });
+        }
     }
 }
 
 fn lower_box_layout(
-    component: &Rc<Component>,
-    rect: LayoutRect,
+    _component: &Rc<Component>,
     layout_element: &ElementRc,
-    collected_children: &mut Vec<ElementRc>,
-    type_register: &TypeRegister,
     style_metrics: &Option<Rc<Component>>,
     diag: &mut BuildDiagnostics,
-) -> Option<Layout> {
-    let is_horizontal = layout_element.borrow().base_type.to_string() == "HorizontalLayout";
+    is_horizontal: bool,
+) {
     let mut layout = BoxLayout {
         is_horizontal,
         elems: Default::default(),
-        geometry: LayoutGeometry::new(rect, &layout_element, style_metrics),
+        geometry: LayoutGeometry::new(&layout_element, style_metrics),
     };
+
+    let layout_cache_prop = create_new_prop(layout_element, "layout_cache", Type::LayoutCache);
+    let layout_info_prop = create_new_prop(layout_element, "layoutinfo", layout_info_type());
+
     let layout_children = std::mem::take(&mut layout_element.borrow_mut().children);
-    for layout_child in layout_children {
-        if let Some(item) = create_layout_item(
-            &layout_child,
-            component,
-            collected_children,
-            type_register,
-            style_metrics,
-            diag,
-        ) {
+    for layout_child in &layout_children {
+        if let Some(item) =
+            create_layout_item(layout_child, layout.elems.len(), &layout_cache_prop, diag)
+        {
             layout.elems.push(item)
         }
     }
-    component.optimized_elements.borrow_mut().push(layout_element.clone());
-    if !layout.elems.is_empty() {
-        Some(layout.into())
-    } else {
-        None
-    }
+    layout_element.borrow_mut().children = layout_children;
+    layout_cache_prop.element().borrow_mut().bindings.insert(
```

Comment:
Perhaps if the `lower_$type_layout()` functions returned a `Layout`, then the binding insertion code for both properties could be moved to the caller side and de-duplicated.

Reply:
Yeah i actually thought about that, but then i also need to pass the layout_cache_prop and layout_info_prop to the function.